hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -0,0 +1,360 @@
+// The parameters source and destination are funcs because AST fragments can't be reused, and in
+// some cases we need to reference source and destination multiple times in a single fragment.
+type StorageTypeConversion func(reader dst.Expr, writer dst.Expr, ctx *CodeGenerationContext) []dst.Stmt
+
+// StorageTypeConversionFactory represents factory methods that can be used to create StorageTypeConversions
+// for a specific pair of types
+// source is the endpoint that will be read
+// destination is the endpoint that will be written
+type StorageTypeConversionFactory func(source *StorageConversionEndpoint, destination *StorageConversionEndpoint) StorageTypeConversion
+
+// A list of all known type conversion factory methods
+var typeConversionFactories []StorageTypeConversionFactory
+
+func init() {
+	typeConversionFactories = []StorageTypeConversionFactory{
+		assignPrimitiveTypeFromPrimitiveType,
+		assignOptionalPrimitiveTypeFromPrimitiveType,
+		assignPrimitiveTypeFromOptionalPrimitiveType,
+		assignArrayFromArray,
+		assignMapFromMap,
+	}
+}
+
+// createTypeConversion tries to create a type conversion between the two provided types, using
+// all of the available type conversion functions in priority order to do so.
+func createTypeConversion(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint) (StorageTypeConversion, error) {
+	for _, f := range typeConversionFactories {
+		result := f(sourceEndpoint, destinationEndpoint)
+		if result != nil {
+			return result, nil
+		}
+	}
+
+	err := fmt.Errorf(
+		"no conversion found to assign %s from %s",
+		destinationEndpoint.name,
+		sourceEndpoint.name)
+
+	return nil, err
+}
+
+// assignPrimitiveTypeFromPrimitiveType will generate a direct assignment if both types have the
+// same underlying primitive type and have the same optionality
+//
+// <destination> = <source>
+//
+func assignPrimitiveTypeFromPrimitiveType(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint) StorageTypeConversion {
+
+	srcOpt := AsOptionalType(sourceEndpoint.Type())
+	dstOpt := AsOptionalType(destinationEndpoint.Type())
+
+	if srcOpt != dstOpt {
+		// Different optionality, handled elsewhere
+		return nil
+	}
+
+	// Don't start from srcOpt or dstOpt in case they're nil
+	srcPrim := AsPrimitiveType(sourceEndpoint.Type())
+	dstPrim := AsPrimitiveType(destinationEndpoint.Type())
+
+	if srcPrim == nil || dstPrim == nil || !srcPrim.Equals(dstPrim) {
+		// Either or both sides are not primitive types, or not the same primitive type
+		return nil
+	}
+
+	return func(reader dst.Expr, writer dst.Expr, ctx *CodeGenerationContext) []dst.Stmt {
+		return []dst.Stmt{
+			astbuilder.SimpleAssignment(writer, token.ASSIGN, reader),
+		}
+	}
+}
+
+// assignOptionalPrimitiveTypeFromPrimitiveType will generate a direct assignment if both types
+// have the same underlying primitive type and have the same optionality
```

Comment:
Comment doesn't match what the method does?

Reply:
Fixed.